File: easy/ts/205.ts
// https://leetcode.com/problems/isomorphic-strings/?envType=study-plan-v2&envId=top-interview-150
// Time: O(n)
// Sapce: O(1)
function isIsomorphic(s: string, t: string): boolean {
  let mapSToT = {};
  let mapTToS = {};

  let sIndex = 0;
  let tIndex = 0;

  while (sIndex < s.length && tIndex < t.length) {
    if (!mapSToT[s[sIndex]] && !mapTToS[t[tIndex]]) {
      mapSToT[s[sIndex]] = t[tIndex];
      mapTToS[t[tIndex]] = s[sIndex];
      sIndex++;
      tIndex++;
    } else if (
      mapSToT[s[sIndex]] === t[tIndex] &&
      mapTToS[t[tIndex]] === s[sIndex]
    ) {
      sIndex++;
      tIndex++;
    } else {
      return false;
    }
  }
  return true;
}

File: easy/python/205.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        map_s_to_t = {}
        map_t_to_s = {}

        s_index = 0
        t_index = 0

        while s_index < len(s) and t_index < len(t):
            if s[s_index] not in map_s_to_t and t[t_index] not in map_t_to_s:
                map_s_to_t[s[s_index]] = t[t_index]
                map_t_to_s[t[t_index]] = s[s_index]
                s_index += 1
                t_index += 1
            elif (
                s[s_index] in map_s_to_t
                and t[t_index] in map_t_to_s
                and map_s_to_t[s[s_index]] == t[t_index]
                and map_t_to_s[t[t_index]] == s[s_index]
            ):
                s_index += 1
                t_index += 1
            else:
                return False
        return True
